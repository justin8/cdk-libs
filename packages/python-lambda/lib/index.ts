import { Construct } from "@aws-cdk/core";
import * as cdk from "@aws-cdk/core";
import {
  FunctionProps,
  LayerVersion,
  RuntimeFamily,
  Function,
  Code,
  AssetCode,
} from "@aws-cdk/aws-lambda";
import { existsSync, mkdirSync } from "fs";
import { execSync } from "child_process";
import { checkServerIdentity } from "tls";

export interface PythonFunctionProps extends Omit<FunctionProps, "code"> {
  /**
   * The source code of your Lambda function. You can point to a file in an
   * Amazon Simple Storage Service (Amazon S3) bucket or specify your source
   * code as inline text.
   *
   * Only Asset based code is supported for the PythonFunction class
   */
  readonly code: AssetCode;

  /**
   * The directory to store temporary builds of the modules from your requirements.txt
   *
   * @default - "build/${name}"
   */
  readonly buildDir?: string;
}

export class PythonFunction extends Function {
  public requirementsLayer: LayerVersion;

  constructor(parent: Construct, name: string, props: PythonFunctionProps) {
    const buildDir = props.buildDir || `build/${name}`;

    if (props.runtime.family != RuntimeFamily.PYTHON) {
      throw new TypeError("Only Python language families are supported");
    }

    if (!existsSync("build")) {
      mkdirSync("build");
    }

    if (!existsSync(buildDir)) {
      mkdirSync(buildDir);
    }

    execSync(
      `docker run --rm` +
        ` --volume "$PWD/${props.code.path}:/src"` +
        ` --volume "$PWD/${buildDir}:/build"` +
        ` --workdir "/src"` +
        ` ${props.runtime.bundlingDockerImage.image}` +
        ` pip install --requirement requirements.txt --target "/build/python"`
    );

    const requirementsLayer = new LayerVersion(parent, `${name}Requirements`, {
      code: Code.fromAsset(buildDir),
      compatibleRuntimes: [props.runtime],
    });

    const lambdaProps = {
      ...props,
      layers: [...(props.layers || []), requirementsLayer],
    };

    super(parent, name, lambdaProps);
  }
}
