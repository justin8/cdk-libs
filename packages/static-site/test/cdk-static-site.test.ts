import {
  expect as expectCDK,
  haveResource,
  SynthUtils,
  haveResourceLike
} from "@aws-cdk/assert";
import cdk = require("@aws-cdk/core");
import { StaticSite } from "../lib/index";
import { PriceClass } from "@aws-cdk/aws-cloudfront";
import { pathToFileURL } from "url";
import path = require("path");

let defaultProps: any = {};

const basicApp = new cdk.App();
const basicStack = new cdk.Stack(basicApp, "BasicStack");
new StaticSite(basicStack, "MyTestConstruct", defaultProps);

// Update snapshot with `npm test -- -u`
// Things to check with the snapshot:
// * Bucket policy is unchanged
// * CFNOutput of the domain name exists
test("Snapshot output is unchanged", () => {
  expect(SynthUtils.toCloudFormation(basicStack)).toMatchSnapshot();
});

test("S3 bucket created", () => {
  expectCDK(basicStack).to(haveResource("AWS::S3::Bucket"));
});

test("OAI created", () => {
  expectCDK(basicStack).to(
    haveResource("AWS::CloudFront::CloudFrontOriginAccessIdentity")
  );
});

test("Distribution created", () => {
  expectCDK(basicStack).to(haveResource("AWS::CloudFront::Distribution"));
});

test("Distribution config is correct", () => {
  expectCDK(basicStack).to(
    haveResourceLike("AWS::CloudFront::Distribution", {
      DistributionConfig: {
        PriceClass: PriceClass.PRICE_CLASS_ALL,
        IPV6Enabled: true,
        HttpVersion: "http2",
        DefaultRootObject: "index.html",
        Origins: [
          {
            DomainName: {
              "Fn::GetAtt": ["MyTestConstruct0062AECC", "RegionalDomainName"]
            }
          }
        ]
      }
    })
  );
});

// deployment function
test("Deployment is not run by default", () => {
  expectCDK(basicStack).notTo(haveResourceLike("Custom::CDKBucketDeployment"));
});

test("Deployment is run when source is specified", () => {
  let localProps = {
    ...defaultProps,
    source: { path: path.resolve("./test/stub-deployment") }
  };
  const basicApp = new cdk.App();
  const basicStack = new cdk.Stack(basicApp, "BasicStack");
  new StaticSite(basicStack, "MyTestConstruct", localProps);

  expectCDK(basicStack).to(haveResourceLike("Custom::CDKBucketDeployment"));
});

// test("Metric Filter group created", () => {
//   expectCDK(basicStack).to(haveResource("AWS::Logs::MetricFilter"));
// });

// test("Errors Alarm created", () => {
//   expectCDK(basicStack).to(
//     haveResource("AWS::CloudWatch::Alarm", {
//       MetricName: "MyTestConstructErrors"
//     })
//   );
// });

// test("No Logs Alarm created", () => {
//   expectCDK(basicStack).to(
//     haveResource("AWS::CloudWatch::Alarm", {
//       MetricName: "IncomingLogEvents",
//       Namespace: "AWS/Logs"
//     })
//   );
// });

// test("No logs alarm can be disabled", () => {
//   let localProps = { ...defaultProps };
//   localProps.noLogsAlarm = { enabled: false };
//   localProps.alarmsTopic = null;
//   const basicApp = new cdk.App();
//   const basicStack = new cdk.Stack(basicApp, "BasicStack");
//   new LogGroupWrapper(basicStack, "MyTestConstruct", localProps);
//   expectCDK(basicStack).notTo(
//     haveResource("AWS::CloudWatch::Alarm", {
//       MetricName: "IncomingLogEvents",
//       Namespace: "AWS/Logs"
//     })
//   );
// });

// test("Errors alarm can be disabled", () => {
//   let localProps = { ...defaultProps };
//   localProps.errorsAlarm = { enabled: false };
//   localProps.alarmsTopic = null;
//   const basicApp = new cdk.App();
//   const basicStack = new cdk.Stack(basicApp, "BasicStack");
//   new LogGroupWrapper(basicStack, "MyTestConstruct", localProps);
//   expectCDK(basicStack).notTo(
//     haveResource("AWS::CloudWatch::Alarm", {
//       MetricName: "MyTestConstructErrors"
//     })
//   );
// });

// test("Alarms are subscribed to given topic", () => {
//   expectCDK(basicStack).to(
//     haveResource("AWS::CloudWatch::Alarm", {
//       MetricName: "IncomingLogEvents",
//       Namespace: "AWS/Logs",
//       AlarmActions: [{ Ref: "topic69831491" }]
//     })
//   );
//   expectCDK(basicStack).to(
//     haveResource("AWS::CloudWatch::Alarm", {
//       MetricName: "MyTestConstructErrors",
//       AlarmActions: [{ Ref: "topic69831491" }]
//     })
//   );
// });

// test("No subscriptions when alarmsTopic is unspecified", () => {
//   let localProps = { ...defaultProps };
//   localProps.alarmsTopic = null;
//   const basicApp = new cdk.App();
//   const basicStack = new cdk.Stack(basicApp, "BasicStack");
//   new LogGroupWrapper(basicStack, "MyTestConstruct", localProps);
//   expectCDK(basicStack).notTo(
//     haveResource("AWS::CloudWatch::Alarm", {
//       AlarmActions: [{ Ref: "topic69831491" }]
//     })
//   );
// });
