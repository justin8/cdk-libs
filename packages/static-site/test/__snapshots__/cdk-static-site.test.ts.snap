// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot output is unchanged 1`] = `
Object {
  "Outputs": Object {
    "MyTestConstructURLF867ECEA": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "MyTestConstructDistributionCFDistribution93AF8D7C",
          "DomainName",
        ],
      },
    },
  },
  "Resources": Object {
    "MyTestConstructDistributionCFDistribution93AF8D7C": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [],
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "MyTestConstructbucketAC9E5A87",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "MyTestConstructOAI3FFF55EF",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_All",
          "ViewerCertificate": Object {
            "CloudFrontDefaultCertificate": true,
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "MyTestConstructOAI3FFF55EF": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "MyTestConstructbucketAC9E5A87": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "MyTestConstructbucketPolicy5AD14DC8": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "MyTestConstructbucketAC9E5A87",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "MyTestConstructOAI3FFF55EF",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "MyTestConstructbucketAC9E5A87",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
